### PR TITLE
refactor(manager/cargo): use `zod` to parse manifest/config

### PR DESCRIPTION
**Pull request by mkniewallner**
## Changes

Move parsing of `cargo` manifest and config to `zod`, as preliminary steps for adding new features.

## Context

Wanted to add support for git dependencies in `cargo`, but it felt like we could benefit from moving the parsing of manifest and config to `zod` before adding new features. I'm not an expert in using `zod` so I hope the changes make sense.

I tried to split the commits in a meaningful way, since the changes are incremental. If you feel that the PR is too big, let me know, I'd be happy to split it.

There's a pretty strong test suite for the different features we support for the manager, so I trust that they show the refactoring does not break any existing feature, but if needed, I can test the changes on a real repository as well, just let me know.

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/manager/cargo/schema.ts [hunk 1]
```diff
@@ -1,5 +1,162 @@
...
+        if (path) {
+          skipReason = 'path-dependency';
+        } else if (git) {
+          skipReason = 'git-dependency';
+        } else if (workspace) {
+          skipReason = 'inherited-dependency';
+        }
+
+        const dep: PackageDependency = {
+          currentValue: currentValue as any,
```
Comment (viceice) [inline lib/modules/manager/cargo/schema.ts:47]:
```suggestion
          currentValue,
```

File: lib/modules/manager/cargo/schema.ts [hunk 2]
```diff
@@ -1,5 +1,162 @@
...
+const CargoSection = z.object({
+  dependencies: CargoDeps.transform((record) => {
+    const deps: PackageDependency[] = [];
+
+    for (const dep of Object.values(record)) {
+      dep.depType = 'dependencies';
+      deps.push(dep);
+    }
+
+    return deps;
```
Comment (viceice) [inline lib/modules/manager/cargo/schema.ts:96]: extract a reusable helper function, so we don't have so much duplicate code.

File: lib/modules/manager/cargo/types.ts [hunk 3]
```diff
@@ -1,53 +1,5 @@
 import type { DEFAULT_REGISTRY_URL } from './utils';
 
-export interface CargoPackage {
-  /** Semver version */
-  version: string;
-}
-
-export interface CargoDep {
-  /** Path on disk to the crate sources */
```
Comment (viceice) [inline lib/modules/manager/cargo/types.ts:9]: please add those comments to the zod schema

File: lib/modules/manager/cargo/extract.ts [hunk 4]
```diff
@@ -223,13 +172,16 @@ export async function extractPackageFile(
...
   const cargoRegistries = extractCargoRegistries(cargoConfig);
 
-  let cargoManifest: CargoManifest;
-  try {
-    cargoManifest = parseToml(content) as CargoManifest;
-  } catch (err) {
+  const { val: cargoManifest, err } = Result.parse(
+    content,
+    CargoManifestSchema,
+  ).unwrap();
```
Comment (viceice) [inline lib/modules/manager/cargo/extract.ts:178]:
```suggestion
  const { val: cargoManifest, err } =  CargoManifestSchema.safeParse(content);
```

simplify

File: lib/modules/manager/cargo/extract.ts [hunk 5]
```diff
@@ -135,12 +81,15 @@ async function readCargoConfig(): Promise<CargoConfig | null> {
...
     const payload = await readLocalFile(path, 'utf8');
     if (payload) {
-      try {
-        return parseToml(payload) as CargoConfig;
-      } catch (err) {
-        logger.debug({ err }, `Error parsing ${path}`);
+      const { val: cargoConfig, err } = Result.parse(
+        payload,
+        CargoConfigSchema,
+      ).unwrap();
```
Comment (viceice) [inline lib/modules/manager/cargo/extract.ts:87]: see other comment and simplify. no need for wrapping

File: lib/modules/manager/cargo/schema.ts [hunk 6]
```diff
@@ -1,5 +1,132 @@
...
+        };
+
+        if (skipReason) {
+          dep.skipReason = skipReason;
+        }
+        if (pkg) {
+          dep.packageName = pkg;
+        }
+        if (registry) {
+          dep.managerData!.registry = registry;
```
Comment (viceice) [inline lib/modules/manager/cargo/schema.ts:65]:
```suggestion
          dep.managerData!.registryName = registry;
```

makes it more clear.

You can also define an interface for `managerData`, so we're more type safe. `PackageDependency` is generic 😉

**Review by rarkins — DISMISSED** [review]

File: lib/modules/manager/cargo/schema.ts [hunk 7]
```diff
@@ -1,5 +1,133 @@
...
+    })
+    .transform(
+      ({
+        path,
+        git,
+        version,
+        registry,
+        package: pkg,
+        workspace,
+      }): PackageDependency => {
```
Comment (viceice) [inline lib/modules/manager/cargo/schema.ts:32]:
```suggestion
      }): PackageDependency<CargoManagerData> => {
```

File: lib/modules/manager/cargo/extract.ts [hunk 8]
```diff
@@ -28,75 +28,32 @@ function getCargoIndexEnv(registryName: string): string | null {
...
-
-    if (typeof currentValue !== 'string') {
-      const version = currentValue.version;
-      const path = currentValue.path;
-      const git = currentValue.git;
-      const registryName = currentValue.registry;
-      const workspace = currentValue.workspace;
 
-      packageName = currentValue.package;
+  const deps: PackageDependency[] = [];
```
Comment (viceice) [inline lib/modules/manager/cargo/extract.ts:39]:
```suggestion
  const deps: PackageDependency<CargoManagerData>[] = [];
```

**Review by viceice — APPROVED** [review]

**Pull request merged**